File: generated_codes/chat_agents/chat-agents-cpp/skiplist/claude_sonnet_4.6/output4_skiplist.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <random>

class ConcurrentDataStructure : public SetADT {
    static constexpr int MAX_LEVEL = 16;

    struct Node {
        int val;
        int topLevel;
        std::atomic<uint64_t> epoch;
        std::atomic<Node*> forward[MAX_LEVEL];

        Node(int v, int lvl) : val(v), topLevel(lvl), epoch(0) {
            for (int i = 0; i < MAX_LEVEL; i++)
                forward[i].store(nullptr, std::memory_order_relaxed);
        }
    };

    static bool is_marked_ref(Node* p) { return (reinterpret_cast<uintptr_t>(p) & 1UL) != 0; }
    static Node* get_unmarked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1UL); }
    static Node* get_marked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1UL); }

    Node* head_;
    Node* tail_;
    std::mt19937 rng_;
    std::uniform_real_distribution<double> dist_;
    std::atomic<uint64_t> globalEpoch_;

    int randomLevel() {
        int lvl = 1;
        while (dist_(rng_) < 0.5 && lvl < MAX_LEVEL) lvl++;
        return lvl;
    }

    bool find(int key, Node** preds, Node** succs) {
    retry:
        Node* pred = head_;
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            Node* curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
            while (true) {
                Node* succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                while (is_marked_ref(succ_raw)) {
                    Node* snext = get_unmarked_ref(succ_raw);
                    Node* expected = curr;
                    if (!pred->forward[lv].compare_exchange_strong(expected, snext,
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        goto retry;
                    curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
                    succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                }
                if (curr->val < key) { pred = curr; curr = get_unmarked_ref(succ_raw); }
                else break;
            }
            preds[lv] = pred;
            succs[lv] = curr;
        }
        return succs[0]->val == key;
    }

public:
    ConcurrentDataStructure() : rng_(std::random_device{}()), dist_(0.0, 1.0), globalEpoch_(0) {
        head_ = new Node(INT_MIN, MAX_LEVEL);
        tail_ = new Node(INT_MAX, MAX_LEVEL);
        for (int i = 0; i < MAX_LEVEL; i++)
            head_->forward[i].store(tail_, std::memory_order_relaxed);
    }

    ~ConcurrentDataStructure() {
        Node* curr = head_;
        while (curr) {
            Node* next = get_unmarked_ref(curr->forward[0].load(std::memory_order_relaxed));
            delete curr;
            curr = next;
        }
    }

    bool contains(int key) override {
        uint64_t ep = globalEpoch_.load(std::memory_order_acquire);
        Node* pred = head_;
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            Node* curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
            while (curr->val < key)
                curr = get_unmarked_ref(curr->forward[lv].load(std::memory_order_acquire));
            pred = (curr->val == key) ? curr : pred;
        }
        Node* candidate = get_unmarked_ref(pred->forward[0].load(std::memory_order_acquire));
        bool found = candidate->val == key
            && !is_marked_ref(candidate->forward[0].load(std::memory_order_acquire));
        (void)ep;
        return found;
    }

    bool add(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        int lvl = randomLevel();
        while (true) {
            if (find(key, preds, succs)) return false;
            Node* newNode = new Node(key, lvl);
            newNode->epoch.store(globalEpoch_.fetch_add(1, std::memory_order_acq_rel),
                std::memory_order_release);
            for (int i = 0; i < lvl; i++)
                newNode->forward[i].store(succs[i], std::memory_order_relaxed);
            Node* expected = succs[0];
            if (!preds[0]->forward[0].compare_exchange_strong(expected, newNode,
                    std::memory_order_acq_rel, std::memory_order_acquire)) {
                delete newNode; continue;
            }
            for (int i = 1; i < lvl; i++) {
                while (true) {
                    expected = succs[i];
                    if (preds[i]->forward[i].compare_exchange_strong(expected, newNode,
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        break;
                    find(key, preds, succs);
                }
            }
            return true;
        }
    }

    bool remove(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        while (true) {
            if (!find(key, preds, succs)) return false;
            Node* target = succs[0];
            for (int i = target->topLevel - 1; i >= 1; i--) {
                Node* succ = target->forward[i].load(std::memory_order_acquire);
                while (!is_marked_ref(succ)) {
                    if (target->forward[i].compare_exchange_weak(succ, get_marked_ref(succ),
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        break;
                }
            }
            Node* succ = target->forward[0].load(std::memory_order_acquire);
            while (true) {
                if (is_marked_ref(succ)) return false;
                if (target->forward[0].compare_exchange_strong(succ, get_marked_ref(succ),
                        std::memory_order_acq_rel, std::memory_order_acquire)) {
                    globalEpoch_.fetch_add(1, std::memory_order_acq_rel);
                    find(key, preds, succs);
                    return true;
                }
            }
        }
    }
};

File: generated_codes/chat_agents/chat-agents-java/skiplist/chatgpt_java/output4_skiplist.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {
private static final int MAX_LEVEL = 32;
private final Node head;
private final Node tail;


private static final class Node {
    final int key;
    final int topLevel;
    final AtomicMarkableReference<Node>[] next;

    @SuppressWarnings("unchecked")
    Node(int key, int topLevel) {
        this.key = key;
        this.topLevel = topLevel;
        this.next = (AtomicMarkableReference<Node>[]) new AtomicMarkableReference[topLevel + 1];
        for (int i = 0; i <= topLevel; i++) {
            this.next[i] = new AtomicMarkableReference<>(null, false);
        }
    }
}

public ConcurrentDataStructure() {
    this.head = new Node(Integer.MIN_VALUE, MAX_LEVEL);
    this.tail = new Node(Integer.MAX_VALUE, MAX_LEVEL);
    for (int i = 0; i <= MAX_LEVEL; i++) {
        head.next[i].set(tail, false);
    }
}

@Override
public boolean add(int key) {
    int topLevel = randomLevel();
    Node[] preds = new Node[MAX_LEVEL + 1];
    Node[] succs = new Node[MAX_LEVEL + 1];

    while (true) {
        if (find(key, preds, succs)) {
            return false;
        }

        Node node = new Node(key, topLevel);
        for (int level = 0; level <= topLevel; level++) {
            node.next[level].set(succs[level], false);
        }

        if (!preds[0].next[0].compareAndSet(succs[0], node, false, false)) {
            continue;
        }

        for (int level = 1; level <= topLevel; level++) {
            while (true) {
                if (preds[level].next[level].compareAndSet(succs[level], node, false, false)) {
                    break;
                }
                find(key, preds, succs);
            }
        }

        return true;
    }
}

@Override
public boolean remove(int key) {
    Node[] preds = new Node[MAX_LEVEL + 1];
    Node[] succs = new Node[MAX_LEVEL + 1];
    boolean[] marked = {false};

    while (true) {
        if (!find(key, preds, succs)) {
            return false;
        }

        Node node = succs[0];

        for (int level = node.topLevel; level >= 1; level--) {
            Node succ = node.next[level].get(marked);
            while (!marked[0]) {
                node.next[level].compareAndSet(succ, succ, false, true);
                succ = node.next[level].get(marked);
            }
        }

        Node succ = node.next[0].get(marked);
        while (true) {
            if (marked[0]) {
                return false;
            }

            if (node.next[0].compareAndSet(succ, succ, false, true)) {
                // Node has been marked
                find(key, preds, succs);
                return true;
            }

            succ = node.next[0].get(marked);
        }
    }
}

@Override
public boolean contains(int key) {
    boolean[] marked = {false};
    Node pred = head;
    Node curr = null;
    Node succ;

    for (int level = MAX_LEVEL; level >= 0; level--) {
        curr = pred.next[level].getReference();

        while (true) {
            succ = curr.next[level].get(marked);

            while (marked[0]) {
                curr = succ;
                succ = curr.next[level].get(marked);
            }

            if (curr.key < key) {
                pred = curr;
                curr = succ;
            } else {
                break;
            }
        }
    }

    return curr.key == key && !curr.next[0].isMarked();
}

private boolean find(int key, Node[] preds, Node[] succs) {
    boolean[] marked = {false};

    retry:
    while (true) {
        Node pred = head;

        for (int level = MAX_LEVEL; level >= 0; level--) {
            Node curr = pred.next[level].getReference();

            while (true) {
                Node succ = curr.next[level].get(marked);

                while (marked[0]) {
                    if (!pred.next[level].compareAndSet(curr, succ, false, false)) {
                        continue retry;
                    }
                    curr = pred.next[level].getReference();
                    succ = curr.next[level].get(marked);
                }

                if (curr.key < key) {
                    pred = curr;
                    curr = succ;
                } else {
                    break;
                }
            }

            preds[level] = pred;
            succs[level] = curr;
        }

        return succs[0].key == key;
    }
}

private int randomLevel() {
    int level = 0;
    while (level < MAX_LEVEL && ThreadLocalRandom.current().nextInt(2) == 0) {
        level++;
    }
    return level;
}


}
